From ab0d3fff7b2284dd562738c8955af23d2cd1a167 Mon Sep 17 00:00:00 2001
From: Jon Turney <jon.turney@dronecode.org.uk>
Date: Fri, 22 May 2026 20:54:31 +0100
Subject: [PATCH 19/19] ld: pep: Stop emitting reference to the
 pseudo-relocator function (v2)

The exported _pei386_runtime_relocator function currently only exists as
a non-functional stub in the Cygwin runtime, merely to satisfy this
reference synthesized in ld [1].

The function still has contents in the MinGW runtime [2], but is already
referenced by being explcitly called at crt startup/dll load.

We'd like to drop the (slightly confusingly named) stub from the Cygwin
runtime, especially in possible future architectures (e.g aarch64),
where backwards compatibilty concerns don't exist.

For all PE/PE+ targets, stop explicitly emitting a reference to
_pei386_runtime_relocator in the linker.

[1] https://cygwin.com/cgit/newlib-cygwin/tree/winsup/cygwin/lib/pseudo-reloc-dummy.c
[2] https://sourceforge.net/p/mingw-w64/mingw-w64/ci/master/tree/mingw-w64-crt/crt/pseudo-reloc.c#l479
---
 ld/pe-dll.c | 53 +----------------------------------------------------
 1 file changed, 1 insertion(+), 52 deletions(-)

diff --git a/ld/pe-dll.c b/ld/pe-dll.c
index 7fd692efb95..fa2f983c1ac 100644
--- a/ld/pe-dll.c
+++ b/ld/pe-dll.c
@@ -170,7 +170,6 @@ static struct bfd_section *edata_s, *reloc_s;
 static unsigned char *edata_d, *reloc_d;
 static unsigned char *reloc_d = NULL;
 static size_t edata_sz, reloc_sz = 0;
-static int runtime_pseudo_relocs_created = 0;
 static bool runtime_pseudp_reloc_v2_init = false;
 
 typedef struct
@@ -1778,7 +1777,7 @@ generate_reloc (bfd *abfd, struct bfd_link_info *info)
   /* This can happen for example when LTO has eliminated all code.  */
   if (total_relocs == 0)
     return;
-  
+
   /* At this point, we have total_relocs relocation addresses in
      reloc_addresses, which are all suitable for the .reloc section.
      We must now create the new sections.  */
@@ -2803,50 +2802,6 @@ make_runtime_pseudo_reloc (const char *name ATTRIBUTE_UNUSED,
   return abfd;
 }
 
-/*	.section	.rdata
-	.rva		__pei386_runtime_relocator  */
-
-static bfd *
-pe_create_runtime_relocator_reference (bfd *parent)
-{
-  asection *extern_rt_rel;
-  unsigned char *extern_rt_rel_d;
-  char *oname;
-  bfd *abfd;
-
-  oname = xasprintf ("%s_ertr%06d.o", dll_symname, tmp_seq);
-  tmp_seq++;
-
-  abfd = bfd_create (oname, parent);
-  free (oname);
-  bfd_find_target (pe_details->object_target, abfd);
-  bfd_make_writable (abfd);
-
-  bfd_set_format (abfd, bfd_object);
-  bfd_set_arch_mach (abfd, pe_details->bfd_arch, 0);
-
-  symptr = 0;
-  symtab = xmalloc (2 * sizeof (asymbol *));
-  extern_rt_rel = quick_section (abfd, ".rdata", SEC_HAS_CONTENTS, 2);
-
-  quick_symbol (abfd, "", U ("_pei386_runtime_relocator"), "", UNDSEC,
-		BSF_NO_FLAGS, 0);
-
-  bfd_set_section_size (extern_rt_rel, PE_IDATA5_SIZE);
-  extern_rt_rel_d = xcalloc (1, PE_IDATA5_SIZE);
-  extern_rt_rel->contents = extern_rt_rel_d;
-
-  quick_reloc (abfd, 0, BFD_RELOC_RVA, 1);
-  save_relocs (extern_rt_rel);
-
-  bfd_set_symtab (abfd, symtab, symptr);
-
-  bfd_set_section_contents (abfd, extern_rt_rel, extern_rt_rel_d, 0, PE_IDATA5_SIZE);
-
-  bfd_make_readable (abfd);
-  return abfd;
-}
-
 #if defined (COFF_WITH_peAArch64)
 static void
 aarch64_make_jump_stub(const char* symbol_name, bfd* parent)
@@ -2974,12 +2929,6 @@ pe_create_import_fixup (arelent *rel, asection *s, bfd_vma addend, char *name,
 				     rel->howto->bitsize,
 				     link_info.output_bfd);
       add_bfd_to_link (b, bfd_get_filename (b), &link_info);
-
-      if (runtime_pseudo_relocs_created++ == 0)
-	{
-	  b = pe_create_runtime_relocator_reference (link_info.output_bfd);
-	  add_bfd_to_link (b, bfd_get_filename (b), &link_info);
-	}
     }
 
   else if (addend != 0)
-- 
2.51.0

